src/main/webapp/user_profile_script.js

```diff
@@ -91,18 +99,88 @@ async function loadMessages(keyString) {
 
     const msgContainer = document.getElementById("message-container");
     msgContainer.innerHTML = "";
-    for (var index = 0; index < msgResponse.length; index++) {
-        var msg = msgResponse[index];
+    if (msgResponse.length > 10) {
+        var msg = msgResponse[msgResponse.length - 10];
         var newMessage = document.createElement("div");
         newMessage.className = msg.className;
         newMessage.appendChild(document.createTextNode(msg.message)); 
+        var moreMessageBtn = document.createElement("button");
+        moreMessageBtn.setAttribute("class", "more-msg-button");
+        moreMessageBtn.textContent = "Load 10 more messages";
+        moreMessageBtn.addEventListener("click", function() { 
+            loadMoreMessages(keyString, newMessage)
+        });
+        msgContainer.appendChild(moreMessageBtn);
         msgContainer.appendChild(newMessage);
+        for (var index = msgResponse.length - 9; index < msgResponse.length; index++) {
+            var message = msgResponse[index];
+            var newMessageDiv = document.createElement("div");
+            newMessageDiv.className = message.className;
+            newMessageDiv.appendChild(document.createTextNode(message.message)); 
+            msgContainer.appendChild(newMessageDiv);
+        }
+    } else {
+        for (var index = 0; index < msgResponse.length; index++) {
+            var msg = msgResponse[index];
+            var newMessage = document.createElement("div");
+            newMessage.className = msg.className;
+            newMessage.appendChild(document.createTextNode(msg.message)); 
+            msgContainer.appendChild(newMessage);
+        }
     }
 
     // Keep message container scrolled to bottom at the beginning
     msgContainer.scrollTop = msgContainer.scrollHeight;
 }
 
+// loadMoreMessages(key, messageDiv) will load 10 more messages of
+// task with keyString equals to key. After loading, scroll to element messageDiv
+async function loadMoreMessages(key, messageDiv) {
+    const queryURL = "/messages?key=" + key;
+    const request = new Request(queryURL, {method: "GET"});
+    const response = await fetch(request);
+    const msgResponse = await response.json();
+
+    const msgContainer = document.getElementById("message-container");
+    msgContainer.removeChild(msgContainer.childNodes[0]);
+    var count = msgContainer.childNodes.length + 10;
```

Comment:
Same comment as above. It might be worth looking into using Cursors in Datastore to reduce the number of message entities retrieved while keeping track of where you are on the query :)

Reply:
Sure. I would use a Cursor entity to keep track of the last retrieved message entity.